### PR TITLE
SyntaxTreeDeserializer: Hold on to omitted nodes across multiple updates

### PR DESCRIPTION
**Pull request by nathawes**
`RawSyntax.init(from decoder: Decoder)` looks up omitted nodes by id from a `WeakLookupTable`, initializing itself as a copy. This copy is then used in the new tree, but wasn't being written back to the lookup table, leaving it with a weak reference to the original node. This was an issue because the `SyntaxTreeDeserializer` only holds a strong reference to the most recent syntax tree, meaning the original node (and table entry) could have been cleaned up, even though further incremental updates could still refer to that id.

Resolves rdar://problem/44877860

### CODE REVIEW

**Comment by nathawes** [comment]
@swift-ci please test

**Comment by nathawes** [comment]
@rintaro I thought it would be better to have the new syntax tree actually use the original RawSyntax reference from the lookup table, but I couldn't work out how to do that...

**Comment by rintaro** [comment]
Perhaps we can use "hackish" factory initializer pattern.
```swift
protocol _RawSyntaxFactory {}
extension _RawSyntaxFactory {
  fileprivate init(_instance: Self) { self = _instance }
}
extension RawSyntax: Codable, _RawSyntaxFactory {
  convenience init(from decoder: Decoder) throws {
    if omitted {
      /* ... */
      let lookupNode = lookupFunc(id)
      self.init(_instance: lookupNode)
      return
    } else {
      /* ... */
    }
  }
}
```

When https://forums.swift.org/t/allow-self-x-in-class-convenience-initializers/15924 is implemented, we probably can migrate to it (i.e. just `self = lookupNode `)

**Comment by nathawes** [comment]
@swift-ci please test

**Comment by nathawes** [comment]
@rintaro Does this look ok now?

File: Sources/SwiftSyntax/RawSyntax.swift [hunk 1]
```diff
@@ -341,7 +341,17 @@ extension RawSyntax {
   }
 }
 
-extension RawSyntax: Codable {
+// This is needed purely to have a self assignment initializer for
+// RawSyntax.init(from:Decoder) so we can reuse omitted nodes, instead of
+// copying them.
+protocol _RawSyntaxFactory {}
```
Comment (rintaro) [inline Sources/SwiftSyntax/RawSyntax.swift]: Sorry, this can be `private`, and we don't need `fileprivate` in `init(_inistance)` decl.

**Comment by nathawes** [comment]
Thanks @rintaro!

**Comment by nathawes** [comment]
@swift-ci please test

**Pull request merged**